napari/layers/_shapes_layer/shape.py

```diff
@@ -0,0 +1,573 @@
+        recongnized
+    _closed : bool
+        Bool if shape edge is a closed path or not
+    _box : np.ndarray
+        9x2 array of vertices of the interaction box. The first 8 points are
+        the corners and midpoints of the box in clockwise order starting in the
+        upper-left corner. The last point is the center of the box
+    _face_vertices : np.ndarray
+        Qx2 array of vertices of all triangles for the shape face
+    _face_triangles : np.ndarray
+        Px3 array of vertex indices that form the triangles for the shape face
+    _edge_vertices : np.ndarray
+        Rx2 array of centers of vertices of triangles for the shape edge.
+        These values should be added to the scaled `_edge_offsets` to get the
+        actual vertex positions. The scaling corresponds to the width of the
+        edge
+    _edge_offsets : np.ndarray
+        Sx2 array of offsets of vertices of triangles for the shape edge. For
+        These values should be scaled and added to the `_edge_vertices` to get
+        the actual vertex positions. The scaling corresponds to the width of
+        the edge
+    _edge_triangles : np.ndarray
+        Tx3 array of vertex indices that form the triangles for the shape edge
+    """
+
+    def __init__(self, *, shape_type='rectangle', edge_width=1,
+                 edge_color='black', face_color='white', opacity=1, z_index=0):
+
+        self._face_vertices = np.empty((0, 2))
+        self._face_triangles = np.empty((0, 3), dtype=np.uint32)
+        self._edge_vertices = np.empty((0, 2))
+        self._edge_offsets = np.empty((0, 2))
+        self._edge_triangles = np.empty((0, 3), dtype=np.uint32)
+        self._box = np.empty((9, 2))
+        self._edge_color_name = 'black'
+        self._face_color_name = 'white'
+
+        self._closed = False
+        self.edge_width = edge_width
+        self.edge_color = edge_color
+        self.face_color = face_color
+        self.opacity = opacity
+        self.z_index = z_index
+
+    @property
+    @abstractmethod
+    def data(self):
+        # user writes own docstring
+        raise NotImplementedError()
+
+    @data.setter
+    @abstractmethod
+    def data(self, data):
+        raise NotImplementedError()
+
+    @property
+    def edge_width(self):
+        """float: thickness of lines and edges.
+        """
+        return self._edge_width
+
+    @edge_width.setter
+    def edge_width(self, edge_width):
+        self._edge_width = edge_width
+
+    @property
+    def edge_color(self):
+        """Color, ColorArray: color of edges
+        """
+        return self._edge_color
+
+    @edge_color.setter
+    def edge_color(self, edge_color):
+        self._edge_color = Color(edge_color)
+        if type(edge_color) is str:
+            self._edge_color_name = edge_color
+        else:
+            rgb = tuple([int(255*x) for x in self._edge_color.rgba[:3]])
+            self._edge_color_name = '#%02x%02x%02x' % rgb
```

Comment:
What is `#%02x%02x%02x`? I'd also define this in some sort of constants file.

Reply:
this is a formatting string to generate a hex string from rgba values. i think i'm going to leave this one here